### PR TITLE
Lde/gemini shplonk in prover

### PR DESCRIPTION
**Pull request by ledwards2225**

### CODE REVIEW

File: cpp/src/aztec/transcript/transcript.cpp [hunk 1]
```diff
@@ -110,18 +111,29 @@ void Transcript::compute_challenge_map()
...
  * transcript interactions must occur sequentially according to the manifest.
+ * Function allows for optional input of circuit_size since this is needed in some
+ * test cases, e.g. instantiating a Sumcheck from a mocked transcript.
  *
  * @param challenge_in
  */
-void Transcript::mock_inputs_prior_to_challenge(const std::string& challenge_in)
+void Transcript::mock_inputs_prior_to_challenge(const std::string& challenge_in, size_t circuit_size)
 {
+    (void)circuit_size;
```
Comment (ledwards2225) [inline cpp/src/aztec/transcript/transcript.cpp]: Delete

File: cpp/src/aztec/honk/pcs/claim.hpp [hunk 2]
```diff
@@ -127,6 +127,11 @@ template <typename Params> struct MLEOpeningClaim {
     using Commitment = typename Params::Commitment;
     using Fr = typename Params::Fr;
 
+    MLEOpeningClaim(auto commitment, auto evaluation)
```
Comment (codygunton) [inline cpp/src/aztec/honk/pcs/claim.hpp:130]: You explained that adding this is necessary to work around an optimization targeting the recursive case.

File: cpp/src/aztec/honk/pcs/commitment_key.test.hpp [hunk 3]
```diff
@@ -23,19 +24,20 @@ namespace {
 constexpr std::string_view kzg_srs_path = "../srs_db/ignition";
 }
 
-template <class CK> inline CK* CreateCommitmentKey();
+template <class CK> inline std::shared_ptr<CK> CreateCommitmentKey();
```
Comment (codygunton) [inline cpp/src/aztec/honk/pcs/commitment_key.test.hpp:27]: IOU on what the right data structure is here, but we should just use this for now.

File: cpp/src/aztec/honk/proof_system/prover.cpp [hunk 4]
```diff
@@ -157,7 +162,7 @@ template <typename settings> void Prover<settings>::compute_grand_product_polyno
 
     // Construct permutation polynomial 'z_perm' in lagrange form as:
     // z_perm = [1 numererator_accum[0][0] numererator_accum[0][1] ... numererator_accum[0][n-2]]
-    polynomial z_perm(proving_key->n, proving_key->n);
+    Polynomial z_perm(proving_key->n, proving_key->n);
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/prover.cpp:165]: polynomial ~> Polynomial is you renaming?

File: cpp/src/aztec/honk/proof_system/prover.cpp [hunk 5]
```diff
@@ -280,7 +285,7 @@ template <typename settings> void Prover<settings>::execute_relation_check_round
 {
     // queue.flush_queue(); // NOTE: Don't remove; we may reinstate the queue
 
-    using Multivariates = sumcheck::Multivariates<barretenberg::fr, waffle::STANDARD_HONK_MANIFEST_SIZE>;
+    using Multivariates = sumcheck::Multivariates<Fr, waffle::TOTAL_NUM_POLYNOMIALS>;
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/prover.cpp]: This should contain the words `HONK` and `STANDARD` in it somewhere.

File: cpp/src/aztec/honk/proof_system/prover.cpp [hunk 6]
```diff
@@ -325,11 +324,58 @@ template <typename settings> void Prover<settings>::execute_relation_check_round
...
+    using MLEOpeningClaim = pcs::MLEOpeningClaim<pcs::kzg::Params>;
+
+    // Construct inputs for Gemini:
+    // - Multivariate opening point u = (u_1, ..., u_d)
+    // - MLE opening claim = {commitment, eval} for each multivariate and shifted multivariate polynomial
+    // - Pointers to multivariate and shifted multivariate polynomials
+    std::vector<Fr> opening_point;
+    std::vector<MLEOpeningClaim> opening_claims;
+    std::vector<MLEOpeningClaim> opening_claims_shifted;
+    std::vector<Polynomial*> multivariate_polynomials;
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/prover.cpp:337]: This is idiomatic to Gemini -- raw pointers are in use there?

File: cpp/src/aztec/honk/proof_system/prover.cpp [hunk 7]
```diff
@@ -343,11 +389,9 @@ template <typename settings> void Prover<settings>::execute_univariatization_rou
  * */
 template <typename settings> void Prover<settings>::execute_pcs_evaluation_round()
 {
-    transcript.apply_fiat_shamir("r");
-    // TODO(Cody): Implement
-    for (size_t round_idx = 0; round_idx < proving_key->log_n; round_idx++) {
-        transcript.add_element("a_" + std::to_string(round_idx), barretenberg::fr(round_idx + 1000).to_buffer());
-    }
+    // TODO(luke): This functionality is performed within Gemini::reduce_prove(), called in the previous round. In the
```
Comment (codygunton) [inline cpp/src/aztec/honk/proof_system/prover.cpp:392]: Right, we discussed this a week or two ago.

File: cpp/src/aztec/honk/sumcheck/polynomials/multivariates.hpp [hunk 8]
```diff
@@ -98,8 +108,10 @@ template <class FF_, size_t num_polys> class Multivariates {
     }
 
     explicit Multivariates(transcript::StandardTranscript transcript)
-        : multivariate_n(
-              static_cast<size_t>(transcript.get_field_element("circuit_size").from_montgomery_form().data[0]))
+        : multivariate_n([](std::vector<uint8_t> buffer) {
```
Comment (codygunton) [inline cpp/src/aztec/honk/sumcheck/polynomials/multivariates.hpp:111]: This is ugly, but it was done in a hurry by me, and I think it's fine for now.

File: cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp [hunk 9]
```diff
@@ -74,18 +55,10 @@ TEST(Sumcheck, Prover)
...
-            transcript::Manifest::RoundManifest({ { .name = "univariate_" + label,
-                                                    .num_bytes = fr_size * honk::StandardHonk::MAX_RELATION_LENGTH,
-                                                    .derived_by_verifier = false } },
-                                                /* challenge_name = */ "u_" + label,
-                                                /* num_challenges_in = */ 1));
-    }
-
-    auto transcript = Transcript(transcript::Manifest(manifest_rounds));
+    // Mock prover-transcript interactions prior to Sumcheck
+    auto transcript = Transcript(StandardHonk::create_unrolled_manifest(0, multivariate_d));
```
Comment (codygunton) [inline cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp:59]: I was a little unsure of this (test more likely to break if it depends on external stuff), but I did this too and I think it's fine for now.

File: cpp/src/aztec/transcript/transcript.cpp (hunk 1 above)
Comment (codygunton) [reply cpp/src/aztec/transcript/transcript.cpp]: :pray:

File: cpp/src/aztec/transcript/transcript.cpp [hunk 10]
```diff
@@ -150,6 +164,10 @@ void Transcript::apply_fiat_shamir(const std::string& challenge_name /*, const b
     // TODO(Cody): Coupling: this line insists that the challenges in the manifest
     // are encountered in the order that matches the order of the proof construction functions.
     // Future architecture should specify this data in a single place (?).
+    // info("apply_fiat_shamir: challenge name match");
```
Comment (codygunton) [inline cpp/src/aztec/transcript/transcript.cpp:166]: Leaving a lot of logging statements in comments. Intended? OK if so.

**Review by codygunton — CHANGES_REQUESTED** [review]
Looks good, mostly just asking questions in the comments.

File: cpp/src/aztec/honk/pcs/claim.hpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/pcs/claim.hpp:130]: Yeah this was so I could construct `MLEOpeningClaim`s in place with emplace_back in the prover. Need a constructor for that and there was previously no need for one.

File: cpp/src/aztec/honk/proof_system/prover.cpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/proof_system/prover.cpp:165]: Yeah they both alias the same thing, just trying to make it consistent throughout the prover and I prefer `Polynomial`

File: cpp/src/aztec/honk/proof_system/prover.cpp (hunk 6 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/proof_system/prover.cpp:337]: Yes, just matching what's done in Gemini. Should probably change in the future

File: cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp (hunk 9 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp:59]: I actually think this is the right thing to do. If we change Sumcheck-related Manifest entries, we must change Sumcheck accordingly. If this test is using some fake Manifest, we then also have to go in and fix that. But this is a Sumcheck test, not a fake Manifest test.

File: cpp/src/aztec/transcript/transcript.cpp (hunk 10 above)
Comment (ledwards2225) [reply cpp/src/aztec/transcript/transcript.cpp:166]: Yeah I wasn't planning on leaving them in the final PR but I'd actually like to while were in major debugging mode. I wind up putting these statements in every branch for debugging anyway.

File: cpp/src/aztec/honk/proof_system/prover.cpp (hunk 5 above)
Comment (ledwards2225) [reply cpp/src/aztec/honk/proof_system/prover.cpp]: Good call, done.

File: cpp/src/aztec/transcript/transcript.cpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/aztec/transcript/transcript.cpp]: done

File: cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp (hunk 9 above)
Comment (codygunton) [reply cpp/src/aztec/honk/sumcheck/sumcheck.test.cpp:59]: OK

File: cpp/src/aztec/honk/proof_system/prover.cpp (hunk 4 above)
Comment (codygunton) [reply cpp/src/aztec/honk/proof_system/prover.cpp:165]: Aha, thanks

**Review by codygunton — APPROVED** [review]

**Pull request merged**